src/nupic/research/temporal_memory.py

```diff
@@ -120,419 +209,192 @@ def compute(self, activeColumns, learn=True):
-    @param connections                 (Connections) Connectivity of layer
+    @param excitedColumn   (dict) Excited Column instance from excitedColumnsGenerator
+    @param prevActiveCells (list)  Active cells in `t-1`
 
     @return (tuple) Contains:
-                      `cell`        (int),
-                      `bestSegment` (int)
+                      `bestSegment`                 (int),
+                      `bestNumActiveSynapses`       (int)
     """
     maxSynapses = 0
-    bestCell = None
-    bestSegment = None
-
-    for cell in cells:
-      segment, numActiveSynapses = self.bestMatchingSegment(
-        cell, activeCells, connections)
-
-      if segment is not None and numActiveSynapses > maxSynapses:
-        maxSynapses = numActiveSynapses
-        bestCell = cell
-        bestSegment = segment
-
-    if bestCell is None:
-      bestCell = self.leastUsedCell(cells, connections)
-
-    return bestCell, bestSegment
-
-
-  def bestMatchingSegment(self, cell, activeCells, connections):
-    """
-    Gets the segment on a cell with the largest number of activate synapses,
-    including all synapses with non-zero permanences.
-
-    @param cell                        (int)         Cell index
-    @param activeCells                 (set)         Indices of active cells
-    @param connections                 (Connections) Connectivity of layer
-
-    @return (tuple) Contains:
-                      `segment`                 (int),
-                      `connectedActiveSynapses` (set)
-    """
-    maxSynapses = self.minThreshold
     bestSegment = None
     bestNumActiveSynapses = None
 
-    for segment in connections.segmentsForCell(cell):
+    for i in excitedColumn["matchingSegments"]:
       numActiveSynapses = 0
 
-      for synapse in connections.synapsesForSegment(segment):
-        synapseData = connections.dataForSynapse(synapse)
-        if ( (synapseData.presynapticCell in activeCells) and
-            synapseData.permanence > 0):
+      for syn in self.connections.synapsesForSegment(self.matchingSegments[i]):
+        synapseData = self.connections.dataForSynapse(syn)
+        if self.binSearch(prevActiveCells, synapseData.presynapticCell) != -1:
           numActiveSynapses += 1
 
       if numActiveSynapses >= maxSynapses:
         maxSynapses = numActiveSynapses
-        bestSegment = segment
+        bestSegment = self.matchingSegments[i]
         bestNumActiveSynapses = numActiveSynapses
 
     return bestSegment, bestNumActiveSynapses
 
 
-  def leastUsedCell(self, cells, connections):
+  def leastUsedCell(self, cells):
     """
     Gets the cell with the smallest number of segments.
     Break ties randomly.
 
-    @param cells       (set)         Indices of cells
-    @param connections (Connections) Connectivity of layer
+    @param cells       (list)         Indices of cells
 
     @return (int) Cell index
     """
     leastUsedCells = set()
```

Comment:
Similarly, this can now be a list, since the cells are always stored in sorted lists.